### PR TITLE
[BEAM-587] take advantage of setup/teardown for KafkaWriter

### PR DESCRIPTION
**Pull request by rangadi**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 1]
```diff
@@ -1324,19 +1324,17 @@ public PDone apply(PCollection<V> input) {
...
+    @Setup
+    public void setup() {
+      if (producerFactoryFnOpt.isPresent()) {
+        producer = producerFactoryFnOpt.get().apply(producerConfig);
+      } else {
+        producer = new KafkaProducer<K, V>(producerConfig);
       }
     }
 
+    // nothing to do in in @StartBundle
```
Comment (dhalperi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: drop

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 2]
```diff
@@ -1348,11 +1346,15 @@ public void processElement(ProcessContext ctx) throws Exception {
...
-    public void finishBundle(Context c) throws Exception {
+    public void finishBundle(Context c) throws IOException {
       producer.flush();
+      checkForFailures();
+    }
+
+    @Teardown
+    public void teardown() {
       producer.close();
       producer = null;
```
Comment (dhalperi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: can drop the `= null` if you like.

Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: done. I was wondering if I should do it or not.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java (hunk 1 above)
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: done.

**Comment by rangadi** [comment]
R: @dhalperi 
This is ready for merge. Travis CI failure looks unrelated. Failed to complete because of a timeout, seems like Travis CI issue.

**Comment by dhalperi** [comment]
LGTM, thanks.

Can you ping the Beam thread where this was an issue, asking them to try again at head?

**Pull request closed without merging**